source/Cosmos.System2/Graphics/Canvas.cs

```diff
@@ -1,1202 +1,1498 @@
-                x2_out = Mode.Columns - 1;
-                y2_out = (Mode.Columns - 1) * m + c;
-            }
-
-            // handle y1
-            if (y1_out < 0)
-            {
-                x1_out = -c / m;
-                y1_out = 0;
-            }
-            else if (y1_out >= Mode.Rows)
-            {
-                x1_out = (Mode.Rows - 1 - c) / m;
-                y1_out = Mode.Rows - 1;
-            }
-
-            // handle y2
-            if (y2_out < 0)
-            {
-                x2_out = -c / m;
-                y2_out = 0;
-            }
-            else if (y2_out >= Mode.Rows)
-            {
-                x2_out = (Mode.Rows - 1 - c) / m;
-                y2_out = Mode.Rows - 1;
-            }
-
-            // final check, to avoid lines that are totally outside bounds
-            if (x1_out < 0 || x1_out >= Mode.Columns || y1_out < 0 || y1_out >= Mode.Rows)
-            {
-                x1_out = 0; x2_out = 0;
-                y1_out = 0; y2_out = 0;
-            }
-
-            if (x2_out < 0 || x2_out >= Mode.Columns || y2_out < 0 || y2_out >= Mode.Rows)
-            {
-                x1_out = 0; x2_out = 0;
-                y1_out = 0; y2_out = 0;
-            }
-
-            // replace inputs with new values
-            x1 = (int)x1_out; y1 = (int)y1_out;
-            x2 = (int)x2_out; y2 = (int)y2_out;
-        }
-
-        /// <summary>
-        /// Calculate new Color from back Color with alpha
-        /// </summary>
-        /// <param name="to">Color to calculate.</param>
-        /// <param name="from">Color used to calculate.</param>
-        /// <param name="alpha">Alpha amount.</param>
-        public Color AlphaBlend(Color to, Color from, byte alpha)
-        {
-            byte R = (byte)((to.R * alpha + from.R * (255 - alpha)) >> 8);
-            byte G = (byte)((to.G * alpha + from.G * (255 - alpha)) >> 8);
-            byte B = (byte)((to.B * alpha + from.B * (255 - alpha)) >> 8);
-            return Color.FromArgb(R, G, B);
-        }
-
-    }
-}
+//#define COSMOSDEBUG
+using System;
+using System.Drawing;
+using System.Collections.Generic;
+using Cosmos.System.Graphics.Fonts;
+
+namespace Cosmos.System.Graphics
+{
+    /// <summary>
+    /// Base class for canvas drawing
+    /// </summary>
+    public abstract class Canvas
+    {
+        /// <summary>
+        /// List of supported video modes
+        /// </summary>
+        public abstract List<Mode> AvailableModes { get; }
+
+        /// <summary>
+        /// Default video mode
+        /// </summary>
+        public abstract Mode DefaultGraphicMode { get; }
+
+        /// <summary>
+        /// Active video mode
+        /// </summary>
+        public abstract Mode Mode { get; set; }
+
+        /// <summary>
+        /// Toggle the use of debugging - potentially faster when disabled
+        /// </summary>
+        public bool SendDebug = true;
+
+        // temporary variables
+        private int arrayX, arrayY;

```

Comment:
Reasoning for this is because of Cosmos Memory Management causing the OS to crash faster when more locals are being used. I didn't plan on this being a permanent solution, just a temporary one until Memory Management is working properly

Reply:
`int` is a value type and is therefor not stored in ram but in the stack. I am also positive that we get simple garbage collection working within a month or two, so going for the permanent solution would be better.